### PR TITLE
Read protobuf lib from environment variable and add compile option /MT

### PR DESCRIPTION
**Pull request by linkerzhang**
Read "PROTOBUF_LIB" from environment variable (in my local case it's "C:\ProgramData\Anaconda3\pkgs\protobuf-3.4.0-py36_vc14_0\Library" after running "conda install -c conda-forge protobuf numpy"), so that the include dirs. and libraries could be relative to "PROTOBUF_LIB".

Adding "/MT" when compiling to fix "MT_STATICRELEASE" and "MD_DYNAMICRELEASE" mismatch.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -93,7 +93,9 @@ def __init__(self):
         super(Protobuf, self).__init__()
         # TODO: allow user specify protobuf include_dirs libraries with flags
         # and environment variables
-        self.libraries = ['protobuf']
+        assert os.getenv('PROTOBUF_LIB')
+        self.libraries = [os.path.join(os.getenv('PROTOBUF_LIB'), "lib\\libprotobuf")]
```
Comment (bddppq) [inline setup.py]: `\` is not cross platform. `os.path.join` can handle this if you split it into separated parts.
Shared library suffixes are different on varies platforms (.e.g `.so`, `.dylib` ...), so it's better to let distutils/setuptools to handle it for you. (i.e. setting `self.libraries`, `self.lib_dirs`and `self.rpaths`).

Comment (linkerzhang) [reply setup.py]: Make sense! Thank you!

Fixed it accordingly.

**Comment by krpopo** [comment]
I confirm that this patch works for me.

File: setup.py [hunk 2]
```diff
@@ -93,7 +94,10 @@ def __init__(self):
         super(Protobuf, self).__init__()
         # TODO: allow user specify protobuf include_dirs libraries with flags
         # and environment variables
-        self.libraries = ['protobuf']
+        if os.getenv('CONDA_PREFIX'):
+            self.libraries = [os.path.join(os.getenv('CONDA_PREFIX'), "Library", "lib", "libprotobuf")]
```
Comment (ezyang) [inline setup.py]: This is not portable. You should test if you're on Windows before doing this.

File: setup.py [hunk 3]
```diff
@@ -185,6 +188,9 @@ def create_extension(ExtType, name, sources, dependencies, extra_link_args, extr
         extra_compile_args.append('-stdlib=libc++')
     if os.getenv('CONDA_PREFIX'):
         include_dirs.append(os.path.join(os.getenv('CONDA_PREFIX'), "include"))
+        include_dirs.append(os.path.join(os.getenv('CONDA_PREFIX'), "Library", "Include"))
```
Comment (ezyang) [inline setup.py]: Ditto here.

Comment (linkerzhang) [reply setup.py]: Sorry, what do you mean "not portable" please? This is similar with the line above, but adding one more include dir which contains protobuf header files.

File: setup.py (hunk 2 above)
Comment (linkerzhang) [reply setup.py]: it works in windows btw :).

Comment (krpopo) [reply setup.py]: It works in Windows, but it breaks for other platforms. If you build in Linux conda environment, `self.libraries` should be `['protobuf']`, not `[os.path.join(os.getenv('CONDA_PREFIX'), "Library", "lib", "libprotobuf")]`.

Your `if` only checks for conda, but you should also check for platform

Comment (linkerzhang) [reply setup.py]: Thank you very much! Limiting it only setting include dir and library for windows only now.

**Comment by ebarsoumMS** [comment]
Hi @ezyang what you mean by don't hardcode Windows, some setup logic is different between Windows and Linux?

**Comment by linkerzhang** [comment]
@ezyang, "/MT" compile option is only needed for windows, I think. Copying /MT description as below,

Causes the application to use the multithread, static version of the run-time library. Defines _MT and causes the compiler to place the library name LIBCMT.lib into the .obj file so that the linker will use LIBCMT.lib to resolve external symbols.

Make sense?

**Comment by ezyang** [comment]
@ebarsoum Do any folks at Microsoft want to setup Windows CI build for onnx? Would be good to prevent this from bitrotting.

@linkerzhang Thank you for this! It all looks good to me.

**Comment by linkerzhang** [comment]
@ezyang Good point! we should set up windows CI build. We'll have someone to do it :).

**Pull request merged**